### ISSUE TITLE
[5] Handle existing OCR/transcripts instead of generating new ones

### ISSUE THREAD

**Issue by orangewolf** [scientist-softserv/louisville-hyku#51]
~~# _(ON HOLD - See comment below)_~~

# ref
https://docs.google.com/document/d/1TfEG3Ztfv2YYcIpxCgPJ2zMq4mJrTHUW/edit?rtpof=true

## story
an imported work with existing ocr text should be used instead of the auto generated text with the newspaper gem

## example csv
[leader_Text_HasChildMetadata.csv](/uploads/706f8b69ecc288e029a491f140878d75/leader_Text_HasChildMetadata.csv)
- the text in the "searchable_text" header is the existing ocr text

## acceptance criteria
- [x] Be able to attach existing OCR data into the records, instead of generating them automatically
- [ ] ~~the ocr doesn't run when a "searchable_text" header is present~~

**Comment by orangewolf** [comment]
Linked Items
---
[EPIC 3 - Handle Existing OCR/Transcripts](https://gitlab.com/notch8/louisville-hyku/-/issues/27)

[Search: Parent/Child Results](https://gitlab.com/notch8/louisville-hyku/-/issues/100)

**Comment by orangewolf** [comment]
Related Merge Requests
---
[i20 - add searchable_text to solr and text_indexer](https://gitlab.com/notch8/louisville-hyku/-/merge_requests/42)

**Comment by orangewolf** [comment]
Shana February 2022
---
I met with Rob and created this [document](https://docs.google.com/document/d/1TfEG3Ztfv2YYcIpxCgPJ2zMq4mJrTHUW/edit), which is more or less a brainstorming session on how to handle existing OCR.
Before moving forward, we need to get clarification from the client.

**Comment by orangewolf** [comment]
Shana February 2022
---
Q:
```
Hi Rachel, 

I have a couple more questions for you regarding OCR. Your reply will dictate how we’ll move forward. 

1. UV (universal viewer) Highlighted Search:
	a.  The universal viewer's search feature enables a user to search for a word and receive highlighted results on the images shown from within the UV. Since the leader text collection does not have coordinate data, we're acknowledging that this feature will be unavailable for it. Because of this, we recommend disabling the UV search across all records and work types for a consistent user experience. Are you in agreement with this? 

To clarify, the user will still be able to search the application for a work's OCR. They just would not be able to do so from within the UV. 

2. Edit OCR: 
	a. I’m curious about your vision for editing OCR. Currently the property “searchable_text” is only available on a Text work type. Is the request to edit OCR on all work types, or just for Texts? 

Thanks!

Shana Moore
```

A:
```
Hi Shana,

We were able to get into the sandbox and will start adding notes/questions to the GitLab boards in preparation for our meeting on Thursday.
 
Here are my reactions to your OCR questions:
For the UV highlighted search, I would prefer to highlight when possible and not worry about a consistent user experience. In other words, we want the best user experience possible for each collection; we actually already have instructions for using the Leader’s clunky search that we can adapt for Hyku to accommodate the differences in the Leader compared to other newspaper collections.
All of the collections that would use OCR will be in the Text work type, and use the searchable_text field.
Thanks,
 
Rachel
```

**Comment by orangewolf** [comment]
Shana February 2022
---
Missed meeting Q: 
```
1) How to handle existing OCR: 
	What does your existing OCR look like? (When reviewing the leader text sample files, is the OCR in the "searchable_text" column or elsewhere?)
	Do you have ALTO files (which contains the coordinates of the OCR words)? This would allow the user to search for the words in the universal viewer, and see the results highlighted.
```

A: 
```
1) How to handle existing OCR: 
	What does your existing OCR look like? (When reviewing the leader text sample files, is the OCR in the "searchable_text" column or elsewhere?)
	Do you have ALTO files (which contains the coordinates of the OCR words)? This would allow the user to search for the words in the universal viewer, and see the results highlighted.
```

**Comment by orangewolf** [comment]
Rachel March 2022
---
Is OCR running at all on staging? I can't successfully search for anything! Example - the name "parrish" should generate results in Simmons, and the name "Juanita" should generate results in Burt.

**Comment by orangewolf** [comment]
Shana March 2022
---
HI @rachel.howard Please don't test anything related to simmons yet. It's not done processing. We kicked off a new import this morning.

You can check its status here. https://lv-hyku-staging.notch8.cloud/importers/32?locale=en Everything in the status column should say complete and have a green check mark, once it's finished.

**Comment by orangewolf** [comment]
Rachel May 2022
---
This does appear to be working! However, I think that the fact that it's working closes ticket #90 rather than this one because this was about searching the full_text field in the metadata when available (it exists for the Leader).

**Comment by orangewolf** [comment]
Alisha May 2022
---
- parent: https://lv-hyku-staging.notch8.cloud/concern/texts/42caa0fc-57ea-494a-8383-a73737b0d799?locale=en
- child: https://lv-hyku-staging.notch8.cloud/concern/texts/26ef6456-3c0e-4770-8fd4-eb46d8ae4503?locale=en
- child's ocr'd text: https://lv-hyku-staging.notch8.cloud/downloads/f8cde689-8a0f-4f5a-bd2b-e5db0665db97?locale=en&file=txt
- child's manually input text (scroll down to the "searchable text" field at the bottom): https://lv-hyku-staging.notch8.cloud/concern/texts/26ef6456-3c0e-4770-8fd4-eb46d8ae4503/edit?locale=en


some of the ocr text does match the "searchable text", but not completely. need to double check this.

**Comment by orangewolf** [comment]
Alisha May 2022
---
@rachel.howard I discussed this with shana and she pointed me towards [this comment](https://gitlab.com/notch8/louisville-hyku/-/issues/20#note_836112312) from a few months ago. 

as is, the catalog search can return results based on text put in the "searchable text" field on a parent or child work. this text however does not have json coordinates, and therefore is not searchable in the uv _unless_ that same term happens to exist in the json/txt files that were created through the ocr process.

we would need the ALTO files (per the comment I referenced above) in order to complete a uv search of the "searchable text".

e.g. using the links I provided above, the "searchable text" and the ocr text of the child include the terms "relationship" and "ku klux". so the catalog search will return the parent work when either of those terms are searched for, and if you click on the child work that contains that text, the uv search will highlight the term.

however, the word "antagonisms" is in the "searchable text" of the child above, but not in the ocr text. searching for that term from the catalog search will return the parent work. while clicking into the first child again (where this searchable text lives), will display "no matches were found" in the uv viewer because "antagonisms" is not in the ocr'd text.

let us know if you have any more questions!

**Comment by orangewolf** [comment]
Rachel May 2022
---
Thanks. There are no ALTO files for this collection (transcription was crowdsourced) and I did expect that highlighting the text on the page would not work for this collection's searchable text. Still, my hope would be that the search would bring them to the child page containing the term if not to the exact place on the page.

**Comment by orangewolf** [comment]
Alisha May 2022
---
> Still, my hope would be that the search would bring them to the child page containing the term if not to the exact place on the page.

you mean that searching a parent uv for a phrase in one of their children's "searchable text" fields will highlight that child's thumbnail in the parent uv correct? and not physically redirect the user to that child work?

**Comment by orangewolf** [comment]
Rachel May 2022
---
Yes, highlighting the thumbnail would be acceptable - it's similar to current CONTENTdm functionality.

<details><summary>...</summary>

From: gitlab@mg.gitlab.com <gitlab@mg.gitlab.com>
Sent: Friday, May 13, 2022 11:38 AM
To: Howard, Rachel <rachel.howard@louisville.edu>
Subject: Re: Louisville-hyku | [5] Handle existing OCR/transcripts instead of generating new ones (#20)

CAUTION: This email originated from outside of our organization. Do not click links, open attachments, or respond unless you recognize the sender's email address and know the contents are safe.

Alisha Evans<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgitlab.com%2Falishaevn&data=05%7C01%7Crachel.howard%40louisville.edu%7Ca4529ec77128475a134b08da34f6921e%7Cdd246e4a54344e158ae391ad9797b209%7C0%7C0%7C637880530872573847%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=PR4%2FH22fCaarBEf0DjuWoafN4ANTY6CusbTI4bE%2BF%2FA%3D&reserved=0> commented on a discussion<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgitlab.com%2Fnotch8%2Flouisville-hyku%2F-%2Fissues%2F20%23note_946265643&data=05%7C01%7Crachel.howard%40louisville.edu%7Ca4529ec77128475a134b08da34f6921e%7Cdd246e4a54344e158ae391ad9797b209%7C0%7C0%7C637880530872573847%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=t3LEemNQU428NPtShhftJk0vNjdAb%2BM5kVkiD%2BPAc10%3D&reserved=0>:

Still, my hope would be that the search would bring them to the child page containing the term if not to the exact place on the page.

you mean that searching a parent uv for a phrase in one of their children's "searchable text" fields will highlight that child's thumbnail in the parent uv correct? and not physically redirect the user to that child work?

-
Reply to this email directly or view it on GitLab<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgitlab.com%2Fnotch8%2Flouisville-hyku%2F-%2Fissues%2F20%23note_946265643&data=05%7C01%7Crachel.howard%40louisville.edu%7Ca4529ec77128475a134b08da34f6921e%7Cdd246e4a54344e158ae391ad9797b209%7C0%7C0%7C637880530872573847%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=t3LEemNQU428NPtShhftJk0vNjdAb%2BM5kVkiD%2BPAc10%3D&reserved=0>.
You're receiving this email because of your account on gitlab.com. If you'd like to receive fewer emails, you can unsubscribe<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgitlab.com%2F-%2Fsent_notifications%2F1145b752e4efb9838534d6a6d142210a%2Funsubscribe&data=05%7C01%7Crachel.howard%40louisville.edu%7Ca4529ec77128475a134b08da34f6921e%7Cdd246e4a54344e158ae391ad9797b209%7C0%7C0%7C637880530872573847%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=b5qYnWBhzHOr1rFCMkAlkoNHp1FPDqWzDfu3vPFo%2FsM%3D&reserved=0> from this thread or adjust your notification settings.

</details>

**Comment by orangewolf** [comment]
Alisha May 2022
---
ok great. can you check off the expectations below that are correct, using the examples below? (when all is said and done, not just right now) am I missing anything?

- parent 1 has both of the children below
- child work 1:
  - ocr txt: "relationship"
  - searchable_text: n/a
- child work 2
  - ocr txt: n/a
  - searchable_text: "antagonisms"

- [x] searching for "relationship" from the catalog search returns parent 1. clicking on parent 1 takes you to its show page where the _term_ "relationship" will be highlighted in the uv, on the child work.
- [x] searching for "antagonisms" from the catalog search returns parent 1. clicking on parent 1 takes you to its show page where no term is highlighted in the uv. however, the _thumbnail_ for child work 2 will be highlighted in the uv viewer. (you may need to click the "contents" sidebar on the right to expand the thumbnails)
- [x] only the uv responds to search terms. the rest of the text, metadata, items, etc. on the parent show page are "static" and will reflect the parent.

**Comment by orangewolf** [comment]
Rachel May 2022
---
The highlighting is not working in the first example (relationship not highlighted in text), although the search does take me to the parent and the child's thumbnail is highlighted in the contents.

**Comment by orangewolf** [comment]
Rachel May 2022
---
I don't seem to be able to check boxes but searching for antagonisms and only uv responds to search terms should be checked.

**Comment by orangewolf** [comment]
Alisha May 2022
---
ok, I've checked the 2nd and 3rd check boxes for you. :slight_smile: 

the list above however was an expectations check, and not a functionality check. would you say that the expectation for check box 1 is correct?

as for functionality -- 
- searching ocr text in a uv: we made some changes that requires a reindex of the data. I'll update you when this is working again
- searching the "searchable text" field in a uv: this is being worked on as part of ticket #100

**Comment by orangewolf** [comment]
Rachel May 2022
---
Thanks for clarifying! This is correct about expectations, although I do wonder if people would find it useful to view the searchable_text field for the child to get a better sense of where the term is. But this situation is unique to this one collection.

**Comment by orangewolf** [comment]
Alisha May 2022
---
ok! I've checked off checkbox 1 as well.

can you elaborate one what you mean by viewing the searchable_text field for a child? (especially since logged out users don't see the child works displayed on the page at all)

**Comment by orangewolf** [comment]
Rachel May 2022
---
This is how it is in CONTENTdm - searchable_text field appears in viewer if selected. I wonder if the UV's "More Description" field could do this but I'm not familiar enough with its customizability to know how what shows up there is determined.


![Image](https://user-images.githubusercontent.com/1054448/203320557-ce6507e7-2601-490b-8d67-68e577d0bbbc.png)

**Comment by orangewolf** [comment]
Alisha May 2022
---
hmmm. I'm not sure what that lift would be either. I'll have to discuss it with the team.

**Comment by orangewolf** [comment]
Alisha May 2022
---
> searching ocr text in a uv: we made some changes that requires a reindex of the data. I'll update you when this is working again

the reindex has completed.

**Comment by orangewolf** [comment]
Alisha May 2022
---
handling of the `searchable_metadata` field in the **parent** uv will be addressed on #100.

refer to [this thread](https://gitlab.com/notch8/louisville-hyku/-/issues/100#note_948482389) for details.

**Comment by orangewolf** [comment]
Shana September 2022
---
I believe this issue can be tested and/or closed along with #100 Since #100 is in QA I will move it to the QA column. cc @kirkkwang 

TLDR copied from this [comment](https://gitlab.com/notch8/louisville-hyku/-/issues/100#note_948482389)

* [x] doing a catalog search for text from the `searchable_text` field of a child should return the parent
* [ ] clicking on that parent should show the uv on the parent show page, and the child(ren) thumbnail(s) that contained the searched for term in its `searchable_text` field should be highlighted
* [x] the "more information" pane on the right side of the uv viewer should contain the complete `searchable_text` field

**Comment by orangewolf** [comment]
Shana September 2022
---
This passes QA: 

User can attach searchable text manually via the work form or a bulkrax import.

[Staging](https://lv-hyku-staging.notch8.cloud/concern/texts/d67e1140-d694-40aa-ba5c-c87ce00118e9?locale=en) - 
1. doing a catalog search for text from the `searchable_text` field of a child should return the parent. when I search for child searchable text "baby cat" the parent gets returned. :white_check_mark: 


![Image](https://user-images.githubusercontent.com/1054448/203321157-d82aa1a9-4c1e-403f-aa66-0e133b907198.png)




2. clicking on that parent should show the uv on the parent show page, and the child(ren) thumbnail(s) that contained the searched for term in its `searchable_text` field should be highlighted => ref #143 

3. the "more information" pane on the right side of the uv viewer should contain the complete `searchable_text` field. :white_check_mark: 



![Image](https://user-images.githubusercontent.com/1054448/203321247-77772870-aee2-4284-83a9-edc510203bee.png)
